File: leetcode/00003_longest_substring_without_repeating_characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        longest = 0
        start = 0
        last_seen = dict()
        for i in range(len(s)):
            ls = last_seen.get(s[i])
            if ls is not None and ls + 1 > start:
                start = ls + 1
            last_seen[s[i]] = i
            longest = max(longest, i - start + 1)    
        return longest

File: leetcode/00003_longest_substring_without_repeating_characters.rs
impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        let chars: Vec<u8> = s.bytes().collect();
        let mut ans = 0;
        let mut left = 0;
        let mut seen = [false; 256];
        for (index, &c) in chars.iter().enumerate() {
            seen[c as usize] ^= true;
            if seen[c as usize] { continue; }
            ans = ans.max(index - left);
            while !seen[c as usize] {
                seen[chars[left] as usize] ^= true;
                left += 1;
            }
        }
        ans.max(chars.len() - left) as i32
    }
}

